### PR TITLE
WIP: Protect against brute force (frequent) login attempts

### PR DESCRIPTION
**Pull request by marefr**
This aims to solve issue #7616 

This introduces a new concept of login attempts. The idea is that we keep a log of invalid login attempts. We use this log when a user in question tries to authenticate to validate if it has a number X (currently 5) of invalid login attempts during the last Y time (currently 5 minutes). 
* If so, we temporary don't allow the user to login. 
* If not and the user in question uses invalid credentials, we add a row to the login attempt log.

There's also a cleanup of old login attempts step added to the cleanup job. Right now this will delete login attempts that are older than 10 minutes.

Right now, login attempts are not captured if ldap is enabled. The reason for not having this if ldap is enabled is that it feels like Grafana instances that has ldap enabled not are exposed to the public internet, but I may be wrong. I would really need some feedback on this.

Questions:
* Do we want/need possibility to configure certain aspects of this feature?
  * Enable/dísable the feature?
  * X, where X is the number of invalid attempts before temporarily blocked?
  * Y, where Y is the historic time stamp we use for counting X

### CODE REVIEW

File: pkg/login/auth.go [hunk 1]
```diff
@@ -2,22 +2,26 @@ package login
...
 	"github.com/grafana/grafana/pkg/bus"
 	m "github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/setting"
 	"github.com/grafana/grafana/pkg/util"
 )
 
 var (
-	ErrInvalidCredentials = errors.New("Invalid Username or Password")
+	ErrInvalidCredentials   = errors.New("Invalid Username or Password")
+	ErrTooManyLoginAttempts = errors.New("Too many consequent incorrect login attempts for user. Login for user temporarily blocked.")
```
Comment (daniellee) [inline pkg/login/auth.go]: Change consequent to consecutive.

File: pkg/login/auth.go [hunk 2]
```diff
@@ -64,3 +80,20 @@ func loginUsingGrafanaDB(query *LoginUserQuery) error {
...
+	loginAttemptCountQuery := m.GetUserLoginAttemptCountQuery{
+		Username: username,
+		Since:    time.Now().Add(time.Minute * -5),
+	}
+
+	if err := bus.Dispatch(&loginAttemptCountQuery); err != nil {
+		return err
+	}
+
+	if loginAttemptCountQuery.Result >= 5 {
```
Comment (daniellee) [inline pkg/login/auth.go]: As discussed, this should probably be configurable.

File: pkg/services/sqlstore/login_attempt.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	return inTransaction(func(sess *DBSession) error {
+		var maxId int64
+		sql := "SELECT max(id) as id FROM login_attempt WHERE created < " + dialect.DateTimeFunc("?")
+		result, err := sess.Query(sql, cmd.OlderThan)
+
+		if err != nil {
+			return err
+		}
+
+		maxId = toInt64(result[0]["id"])
```
Comment (daniellee) [inline pkg/services/sqlstore/login_attempt.go]: To skip this conversion code, you could try a different xorm method than sess.Query. You could try using Find or Get together with a Where clause.

**Comment by bergquist** [comment]
> Right now, login attempts are not captured if ldap is enabled. The reason for not having this if ldap is enabled is that it feels like Grafana instances that has ldap enabled not are exposed to the public internet, but I may be wrong. I would really need some feedback on this.

I do remember someone reporting having all their accounts betting blocked by the LDAP server when exposed to a brute force attack. I have no strong opinions about enable/disable it for public installs but I think some might use it. Not sure we should support it. 

I'm not sure we should make the number of attempts, the time window or how long they are stored configurable. I think we can find sensible defaults instead. Convention over configuration/Less is more etc :)

How about using Exponential back off instead? ex https://support.citrix.com/article/CTX136565
We could allow up to 3 failed attempts the last 1h and then start enforcing a delay until the same IP can try again.

**Comment by marefr** [comment]
> I do remember someone reporting having all their accounts betting blocked by the LDAP server when exposed to a brute force attack. I have no strong opinions about enable/disable it for public installs but I think some might use it. Not sure we should support it.

I think you may refer to this issue: #9900 regarding ldap and brute force attack.

> I'm not sure we should make the number of attempts, the time window or how long they are stored configurable. I think we can find sensible defaults instead. Convention over configuration/Less is more etc :)

Yeah I can definitely see the point of sensible defaults :+1:. Still good to be able to disable the feature though if there should be any problem for an example.

> How about using Exponential back off instead? ex https://support.citrix.com/article/CTX136565
We could allow up to 3 failed attempts the last 1h and then start enforcing a delay until the same IP can try again.

Interesting. Are you thinking of implementing this per username and ip address or only ip address?

**Comment by bergquist** [comment]
> Yeah I can definitely see the point of sensible defaults :+1:. Still good to be able to disable the feature though if there should be any problem for an example.

:100: 

> Interesting. Are you thinking of implementing this per username and ip address or only ip address?

Excellent question. I guess both?

**Comment by torkelo** [comment]
I think this looks very close to mergable. 

- Think we can easily make this apply to LDAP logins as well (just record failed attempts for ldap as well)

Question is what should be configurable and what should the defaults be.

Two options with defaults like?
- login_attempts_per_minute = 3
- login_cool_down = 10m

The look is (currently if I read the code correctly) based on username + IP. Think this is fine for an initial version. 

Thoughts?

**Comment by bergquist** [comment]
I think that sounds like a good start. If people need better protection we can improve this later.

**Comment by marefr** [comment]
> * Think we can easily make this apply to LDAP logins as well (just record failed attempts for ldap as well)

Agree

> The look is (currently if I read the code correctly) based on username + IP. Think this is fine for an initial version.

That's not quite correct. The IP address are currently only stored and not used in login attempt count, only username is. 

Didn't quite see the win of using both username and IP address for locking accounts. Issues regarding IP addresses, i.e. possibility to block certain IP addresses/subnets from accessing Grafana feels like something that you usually use a firewall for. 

Having the IP addresses stored in the login_attempts table can however be helpful to manually verify if some specific IP address are trying to brute force login and then block this in firewall. 

> Two options with defaults like?
> * login_attempts_per_minute = 3
> * login_cool_down = 10m

What I tried to describe above was that we have X and Y in the current implementation. 
* X is the number of invalid attempts before temporarily blocked
* Y is the historic time stamp we use for counting X

So what you refer to as *login_attempts_per_minute* is basically what X and Y is. To support *login_cool_down* I think we need to store the actual login attempts as a counter instead, together with a timestamp. This will allow us to have an account locked for X minutes since the last invalid login attempt that made the counter go above the *max allowed invalid attempts*.

As a comparison, the current implementation works like a sliding window. Example:
* 13:00:00 - invalid login attempt
* 13:01:00 - invalid login attempt
* 13:02:00 - invalid login attempt
* 13:03:00 - invalid login attempt
* 13:04:00 - invalid login attempt
* 13:04:05 - login attempt blocked (counter = 5, account locked)
* 13:05:01 - invalid login attempt (counter = 4)
* 13:05:02 - login attempt blocked (counter = 5, account locked)

**Comment by marefr** [comment]
After discussion with @torkelo:

* We add a possibility to to disable this feature by configuration. It will be enabled per default.
* Make this feature apply to LDAP logins as well.
* Skip checking IP address for now, only store it.
* Some documentation, at least for the configuration parameter.

**Comment by bergquist** [comment]
Sounds great! 

One minor detail. I think it's better to have the default value of a setting be false since its the default value for bools in golang. 

So instead of having `enabled = true` as default. use `disabled = false`
This decreases the chances for bugs due to invalid default values of settings. 

@torkelo @daniellee opinions? 

This is not a rule that we follow in Grafana but I think we should start doing it.

**Comment by marefr** [comment]
Are you meaning that this feature should be disabled per default in configuration or are you referring to the naming of variables, i.e. feature_enabled vs feature_disabled?

**Comment by bergquist** [comment]
I'm referring to the variable name. so `feature_disable = false` should be the default for the setting.

File: pkg/login/auth.go [hunk 4]
```diff
@@ -64,3 +82,20 @@ func loginUsingGrafanaDB(query *LoginUserQuery) error {
 	query.User = user
 	return nil
 }
+
+func validateLoginAttemptCount(username string) error {
+	loginAttemptCountQuery := m.GetUserLoginAttemptCountQuery{
+		Username: username,
+		Since:    time.Now().Add(time.Minute * -5),
```
Comment (eripe970) [inline pkg/login/auth.go]: Should it be possible to configure this as well?

Comment (marefr) [reply pkg/login/auth.go]: No, not in the initial implementation

File: conf/defaults.ini [hunk 5]
```diff
@@ -174,6 +174,9 @@ disable_gravatar = false
 # data source proxy whitelist (ip_or_domain:port separated by spaces)
 data_source_proxy_whitelist =
 
+# disable protection against brute force (frequent) login attempts
+disable_login_attempts_validation = false
```
Comment (marefr) [inline conf/defaults.ini]: I don't know if this a good name

File: pkg/setting/setting.go [hunk 6]
```diff
@@ -75,13 +75,14 @@ var (
...
-	EmailCodeValidMinutes int
-	DataProxyWhiteList    map[string]bool
+	SecretKey                      string
+	LogInRememberDays              int
+	CookieUserName                 string
+	CookieRememberName             string
+	DisableGravatar                bool
+	EmailCodeValidMinutes          int
+	DataProxyWhiteList             map[string]bool
+	DisableLoginAttemptsValidation bool
```
Comment (marefr) [inline pkg/setting/setting.go]: I don't know if this a good name

**Review by marefr — COMMENTED** [review]
A bit unsure of the naming of configuration and setting variable

File: conf/defaults.ini (hunk 5 above)
Comment (torkelo) [reply conf/defaults.ini]: `disable_brute_force_login_protection` seems more clear

Comment (marefr) [reply conf/defaults.ini]: Done. Thanks

File: pkg/api/login.go [hunk 7]
```diff
@@ -102,12 +102,13 @@ func LoginPost(c *middleware.Context, cmd dtos.LoginCommand) Response {
 	}
 
 	authQuery := login.LoginUserQuery{
-		Username: cmd.User,
-		Password: cmd.Password,
+		Username:  cmd.User,
+		Password:  cmd.Password,
+		IpAddress: c.Req.RemoteAddr,
```
Comment (eripe970) [inline pkg/api/login.go]: Will this pick up up the `X-Forwarded-For` header if available?

Comment (torkelo) [reply pkg/api/login.go]: it should (I hope so at least)

Comment (marefr) [reply pkg/api/login.go]: According to macaron source code it does: https://github.com/Unknwon/macaron/blob/master/context.go#L133

**Comment by faxm0dem** [comment]
This is still missing from the online documentation

**Comment by vlada-jerkovic** [comment]
What are you thinking about implementing 3 step Login? This approach has Amazon and works pretty well.
1. Login page with only **username field and Submit button** -> redirect to
2. Page with Security check with **image, input field and Submit button** -> redirect to
3. Page with **Password field and Submit button**
Example:
image

**Pull request merged**